File: strategy/marvin/group/moves.lua
local Moves = Class("Group.Moves")

local debug = require "../base/debug"

function Moves:init()
	self.name = "moves"
	self.moveList = {
		require "group/move/kickoff",
		require "group/move/kickoffdefensive",
		require "group/move/mrltestcorner",
		require "group/move/armada",
		-- require "group/move/overchip",
	}

	for _,move in ipairs(self.moveList) do
		if not move.MIN_ROBOTS or move.MIN_ROBOTS < 0
			or not move.MAX_ROBOTS or move.MAX_ROBOTS < move.MIN_ROBOTS then
			error("MIN_ROBOTS and/or MAX_ROBOT are invalid or not set!")
		end
	end

	self._numAttackersSent = false
	self._chosenMove = nil
	self._currentMove = nil
	self._participatingRobots = {}
end

local function validateAssignment(assignment)
	-- don't assing a task and a behavior
	if assignment.behavior and assignment.class then
		return false
	end

	--don't assing nothing
	if not assignment.behavior and not assignment.class then
		return false
	end

	return true
end

function Moves:run(sender, inbox, messages)
	-- check if all participating robots are still available
	if self._currentMove then
		for _,r in ipairs(self._participatingRobots) do
			if not messages[r] then
				self._currentMove = nil
				self._chosenMove = nil
				break
			end
		end
	end

	-- check if current move can be continued
	if self._currentMove and not self._currentMove:_canContinue() then
		self._currentMove = nil
		self._chosenMove = nil
		self._numAttackersSent = false
	end

	local n_attackers
	-- choose a new move
	if not self._chosenMove then
		local candidates = {}
		local numCandidateRobots = 0
		for _,_ in pairs(inbox.attackerFlag()) do
			numCandidateRobots = numCandidateRobots + 1
		end
		for _,_ in pairs(inbox.defenderFlag()) do
			numCandidateRobots = numCandidateRobots + 1
		end
		for _,move in ipairs(self.moveList) do
			if move.canStart() then
				if numCandidateRobots >= move.MIN_ROBOTS then
					table.insert(candidates, move)
				end
			end
		end

		if #candidates > 0 then
			local index = math.random(#candidates)
			self._chosenMove = candidates[index]
			n_attackers = math.min(numCandidateRobots, candidates[index].MAX_ROBOTS)
		end
	end

	if not self._currentMove and self._chosenMove then
		local availableRobots = {}
		for r,_ in pairs(messages) do
			table.insert(availableRobots, r)
		end

		if #availableRobots >= self._chosenMove.MIN_ROBOTS and
			#availableRobots <= self._chosenMove.MAX_ROBOTS and
			self._numAttackersSent then
			table.sort(availableRobots, function(a, b) return a.id < b.id end)
			self._currentMove = self._chosenMove(availableRobots, inbox)
			self._participatingRobots = availableRobots
		end
	end


	-- reset participating robots
	local prevParticipatingRobots = self._participatingRobots
	self._participatingRobots = {}

	-- run
	if self._currentMove then
		local taskAssignments, mainAttacker = self._currentMove:updateTasks()
		for _, robot in ipairs(prevParticipatingRobots) do
			local assignment = taskAssignments[robot]
			if assignment then
				if not validateAssignment(assignment) then
					for key,value in pairs(assignment) do
						log(tostring(key) .. " -> " .. tostring(value))
					end
					error("invalid assignment for robot " .. tostring(robot.id))
				end
				assignment.mainAttacker = robot == mainAttacker
				sender.moveAssignment(robot, assignment)
				table.insert(self._participatingRobots, robot)
			else
				sender.forcePoolChange("trainer", { robot = robot, destPool = "defender" })
			end
		end
		n_attackers = #self._participatingRobots
	end

	if self._chosenMove and n_attackers then
		assert(n_attackers)
		self._numAttackersSent = true
		sender.moveNumAttackers("trainer", n_attackers)
	end

	debug.push("Move")
	debug.set("ParticipatingRobots", self._participatingRobots)
	if self._currentMove then
		debug.set(nil, Class.name(self._currentMove, true))
	end
	debug.pop()
end

return Moves

File: strategy/typescript/glados/group/moves.ts
import * as debug from "base/debug";
import * as MathUtil from "base/mathutil";
import { FriendlyRobot } from "base/robot";

import { MessageBox, MessageType } from "glados/control/messaging";
import { Armada } from "glados/group/move/armada";
import { BallPlacement } from "glados/group/move/ballplacement";
import { Move } from "glados/group/move/base";
import { CrossShoot } from "glados/group/move/crossshoot";
import { KickOff } from "glados/group/move/kickoff";
import { KickOffDefensive } from "glados/group/move/kickoffdefensive";
import { MrlTestCorner } from "glados/group/move/mrltestcorner";
import { None } from "glados/group/move/none";
import { WindshieldWiper } from "glados/group/move/windshieldwiper";
import { Group } from "glados/trainer/groups";
// import {OverChip} from "glados/group/move/overchip";

export interface Assignment {
	behavior?: any;
	class?: any;
	/** The parameters used to construct class or behavior */
	params: any;
	restart: boolean;
	mainAttacker: boolean;
}

export interface MoveInfo {
	attackers: (number | undefined)[];
	allowExtraAttackers: boolean;
}

export class Moves implements Group {
	public readonly name = "moves";
	private moveList: typeof Move[];
	private _numAttackersSent: boolean = false;
	private _chosenMove: undefined | {
		ctor: typeof Move;
		requestedRobots: number;
	};
	private _currentMove: Move | undefined;
	private _participatingRobots: FriendlyRobot[] = [];

	public constructor(moveList?: typeof Move[]) {
		this.moveList = moveList ?? [
			KickOff,
			KickOffDefensive,
			MrlTestCorner,
			Armada,
			BallPlacement,
			WindshieldWiper,
			CrossShoot,
			None
		];

		for (let move of this.moveList) {
			if (move.MIN_ROBOTS > move.MAX_ROBOTS) {
				throw new Error("Move: MIN_ROBOTS can't be greater than MAX_ROBOTS!");
			}
			if (move.MIN_ROBOTS < 0) {
				throw new Error("Move: MIN_ROBOTS must be greater than or equal to 0!");
			}
		}
	}

	public run(messaging: MessageBox, messages: Map<FriendlyRobot, undefined>) {
		// check if all participating robots are still available
		if (this._currentMove) {
			for (let r of this._participatingRobots) {
				if (!messages.has(r)) {
					this._currentMove = undefined;
					this._chosenMove = undefined;
					this._numAttackersSent = false;
					break;
				}
			}
		}

		// check if current move can be continued
		if (this._currentMove && !this._currentMove.canContinue()) {
			this._currentMove = undefined;
			this._chosenMove = undefined;
			this._numAttackersSent = false;
		}

		let n_attackers;
		let attackers: (undefined | number)[];
		// choose a new move
		if (this._chosenMove == undefined) {
			let candidates = [];
			let numCandidateRobots = 0;
			numCandidateRobots += messaging.receive(MessageType.attackerFlag).size;
			numCandidateRobots += messaging.receive(MessageType.defenderFlag).size;
			for (let move of this.moveList) {
				if (numCandidateRobots >= move.MIN_ROBOTS && move.canStart(numCandidateRobots)) {
					candidates.push(move);
				}
			}

			if (candidates.length > 0) {
				let index = MathUtil.randomInt([0, candidates.length - 1]);
				const chosenMove = candidates[index];
				n_attackers = Math.min(numCandidateRobots, chosenMove.wantedMaxRobots(numCandidateRobots));
				attackers = [];
				attackers.length = n_attackers;

				this._chosenMove = {
					ctor: chosenMove,
					requestedRobots: n_attackers,
				};
			}
		}

		if (this._currentMove == undefined && this._chosenMove != undefined) {
			const availableRobots = [...messages.keys()];

			if (availableRobots.length >= this._chosenMove.ctor.MIN_ROBOTS
					&& this._numAttackersSent) {
				availableRobots.sort((a, b) => b.pos.y - a.pos.y);
				const numAssigned = Math.min(
					availableRobots.length,
					this._chosenMove.requestedRobots,
				);
				this._participatingRobots = availableRobots.slice(0, numAssigned);
				this._currentMove = new (this._chosenMove.ctor as any)(this._participatingRobots, messaging);
			}
		}


		// reset participating robots
		let prevParticipatingRobots = this._participatingRobots;
		this._participatingRobots = [];

		// run
		if (this._currentMove) {
			debug.push("Move");
			let moveParams = this._currentMove.updateTasks();
			for (let robot of prevParticipatingRobots) {
				let assignment: any = moveParams.assignments.get(robot);
				if (assignment != undefined) {
					assignment.mainAttacker = robot === moveParams.mainAttacker;
					if ((assignment.class != undefined && assignment.class !== "none") || assignment.behavior != undefined) {
						messaging.send(MessageType.moveAssignment, robot, assignment);
					}
					this._participatingRobots.push(robot);
				} else {
					messaging.sendToTrainerRepeated(MessageType.forcePoolChange, { robot: robot, destPool: "defender" });
				}
			}
			attackers = this._participatingRobots.map((x) => x.id);
			n_attackers = attackers.length;
			debug.pop();
		}

		if (this._chosenMove && n_attackers != undefined) {
			this._numAttackersSent = true;
			messaging.sendBroadcast(MessageType.moveInfo, {
				attackers: attackers!,
				allowExtraAttackers: this._chosenMove.ctor.ALLOW_EXTRA_ATTACKERS
			});
		}

		debug.push("Move");
		debug.set("ParticipatingRobots", this._participatingRobots);
		if (this._currentMove) {
			debug.set(undefined, this._currentMove.constructor.name);
		}
		debug.pop();
	}
}
